### PR TITLE
feat(genQa): RGA ua events for insight use case created

### PR DESCRIPTION
**Pull request by mmitiche**
[SFINT-5198](https://coveord.atlassian.net/browse/SFINT-5198)

#### Custom events:

- likeGeneratedAnswer
- dislikeGeneratedAnswer
- openGeneratedAnswerSource
- generatedAnswerStreamEnd
- generatedAnswerSourceHover
- generatedAnswerCopyToClipboard
- generatedAnswerHideAnswers
- generatedAnswerShowAnswers
- generativeQuestionFeedbackSubmit

#### Search events:

- retryGeneratedAnswer
- rephraseGeneratedAnswer

### Added the ability to send generated answer metadata in all the Search UA events by adding a new function `getGeneratedAnswerMetadata` to the `InsightClientProvider`

[SFINT-5198]: https://coveord.atlassian.net/browse/SFINT-5198?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/insight/insightClient.ts [hunk 1]
```diff
@@ -476,6 +483,121 @@ export class CoveoInsightClient {
...
+    public logGeneratedAnswerShowAnswers(generatedAnswerMetadata: GeneratedAnswerBaseMeta, metadata?: CaseMetadata) {
+        return this.logCustomEvent(
+            SearchPageEvents.generatedAnswerShowAnswers,
+            metadata
+                ? {...generateMetadataToSend(metadata, false), ...generatedAnswerMetadata}
+                : generatedAnswerMetadata
+        );
+    }
+
+    public logGenerativeQuestionFeedbackSubmit(
```
Comment (erocheleau) [inline src/insight/insightClient.ts:558]:
```suggestion
    public logGeneratedAnswerFeedbackSubmit(
```
???

File: src/insight/insightClient.ts [hunk 2]
```diff
@@ -476,6 +483,121 @@ export class CoveoInsightClient {
...
+        generatedAnswerFeedbackMetadata: GeneratedAnswerFeedbackMeta,
+        metadata?: CaseMetadata
+    ) {
+        return this.logCustomEvent(
+            SearchPageEvents.generativeQuestionFeedbackSubmit,
```
Comment (erocheleau) [inline src/insight/insightClient.ts:563]: ugh we have only this one who is not named like everything else? :(

Comment (erocheleau) [reply src/insight/insightClient.ts:563]: If that's the case I prefer at least the `logX` methods use all the right name and hide this discrepancy .

File: src/insight/insightClient.ts [hunk 3]
```diff
@@ -476,6 +483,121 @@ export class CoveoInsightClient {
...
+            metadata
+                ? {...generateMetadataToSend(metadata, false), ...generatedAnswerFeedbackMetadata}
+                : generatedAnswerFeedbackMetadata
+        );
+    }
+
+    public logRephraseGeneratedAnswer(
```
Comment (erocheleau) [inline src/insight/insightClient.ts:570]:
```suggestion
    public logGeneratedAnswerRephrase(
```
To fit with all the others?

File: src/insight/insightClient.ts [hunk 4]
```diff
@@ -476,6 +483,121 @@ export class CoveoInsightClient {
...
+    ) {
+        return this.logSearchEvent(
+            SearchPageEvents.rephraseGeneratedAnswer,
+            metadata
+                ? {...generateMetadataToSend(metadata, false), ...generatedAnswerRephraseMetadata}
+                : generatedAnswerRephraseMetadata
+        );
+    }
+
+    public logRetryGeneratedAnswer(metadata?: CaseMetadata) {
```
Comment (erocheleau) [inline src/insight/insightClient.ts:582]:
```suggestion
    public logGeneratedAnswerRetry(metadata?: CaseMetadata) {
```
?

**Review by erocheleau — COMMENTED** [review]
Some suggest names change, did we mess up those names in the previous PR ? :(

File: src/insight/insightClient.spec.ts [hunk 5]
```diff
@@ -509,6 +512,116 @@ describe('InsightClient', () => {
...
+        it('should send proper payload for #generatedAnswerShowAnswers', async () => {
+            const exampleGeneratedAnswerMetadata = {
+                generativeQuestionAnsweringId: '123',
+            };
+
+            await client.logGeneratedAnswerShowAnswers(exampleGeneratedAnswerMetadata);
+            expectMatchCustomEventPayload(SearchPageEvents.generatedAnswerShowAnswers, exampleGeneratedAnswerMetadata);
+        });
+
+        it('should send proper payload for #generativeQuestionFeedbackSubmit', async () => {
```
Comment (SimonMilord) [inline src/insight/insightClient.spec.ts:587]:
```suggestion
        it('should send proper payload for #generatedAnswerFeedbackSubmit', async () => {
```

should probably change these if we go with @erocheleau suggestions

File: src/insight/insightClient.spec.ts [hunk 6]
```diff
@@ -509,6 +512,116 @@ describe('InsightClient', () => {
...
+            const exampleGeneratedAnswerMetadata = {
+                generativeQuestionAnsweringId: '123',
+                reason: <GeneratedAnswerFeedbackReason>'other',
+                details: 'foo',
+            };
+
+            await client.logGenerativeQuestionFeedbackSubmit(exampleGeneratedAnswerMetadata);
```
Comment (SimonMilord) [inline src/insight/insightClient.spec.ts:594]: same here

**Review by SimonMilord — APPROVED** [review]
I agree with etienne's suggestions and if we decide to go with that naming then tests naming should also be updated. Other than that it looks fine to me

**Comment by mmitiche** [comment]
> Some suggest names change, did we mess up those names in the previous PR ? :(

yep these were the name we used in the previous PR and the names that were mentioned in the design doc 🤷‍♂️ do we still wanna change ? x)

File: src/insight/insightClient.ts (hunk 1 above)
Comment (mmitiche) [reply src/insight/insightClient.ts]: This event has been updated for both the search and insight use cases

File: src/insight/insightClient.ts (hunk 3 above)
Comment (mmitiche) [reply src/insight/insightClient.ts:570]: I don't think we have a rule about that, looking at all the events we currently have we are using these two patterns interchangeably, thus I don't think this change is worth the trouble especially that  both `logRephraseGeneratedAnswer` and `logGeneratedAnswerRephrase` make sense.

**Review by erocheleau — APPROVED** [review]

**Pull request merged**